File: src/mysqlcookbook/recipes/api/04_stmt/stmt.rb
#!/usr/bin/ruby
# stmt.rb: demonstrate statement processing in Ruby
# (without placeholders)

require "Cookbook"

begin
  client = Cookbook.connect
rescue Mysql2::Error => e
  puts "Cannot connect to server"
  puts "Error code: #{e.errno}"
  puts "Error message: #{e.message}"
  exit(1)
end

puts "Fetch using each as iterator (access columns by position)"
begin
#@ _FETCHLOOP_EACH_ITERATOR_
  result = client.query("SELECT id, name, cats FROM profile")
  result.each do |row|
    printf "id: %s, name: %s, cats: %s\n", row.values[0], row.values[1], row.values[2]
  end
  result.free
#@ _FETCHLOOP_EACH_ITERATOR_
rescue Mysql2::Error => e
  puts "Oops, the statement failed"
  puts "Error code: #{e.errno}"
  puts "Error message: #{e.message}"
end

puts "Fetch using each as iterator (access columns by name)"
begin
  result = client.query("SELECT id, name, cats FROM profile")
#@ _ACCESS_BY_NAME_
  result.each do |row|
    printf "id: %s, name: %s, cats: %s\n",
           row["id"], row["name"], row["cats"]
  end
  result.free
#@ _ACCESS_BY_NAME_
rescue Mysql2::Error => e
  puts "Oops, the statement failed"
  puts "Error code: #{e.errno}"
  puts "Error message: #{e.message}"
end

puts "Execute UPDATE statement"
begin
#@ _DO_1_
  count = client.query("UPDATE profile SET cats = cats+1 WHERE name = 'Sybil'")
  puts "Number of rows updated: #{client.affected_rows}"
#@ _DO_1_
rescue Mysql2::Error => e
  puts "Oops, the statement failed"
  puts "Error code: #{e.errno}"
  puts "Error message: #{e.message}"
end

client.close

File: src/mysqlcookbook/recipes/api/04_stmt/stmt.py
#!/usr/bin/python3
# stmt.py: demonstrate statement processing in Python
# (without placeholders)

import sys
import mysql.connector
import cookbook

try:
  conn = cookbook.connect()
except mysql.connector.Error as e:
  print("Cannot connect to server")
  print("Error code: %s" % e.errno)
  print("Error message: %s" % e.msg)
  sys.exit(1)

print("Fetch rows with fetchone")
try:
#@ _FETCHONE_
  cursor = conn.cursor()
  cursor.execute("SELECT id, name, cats FROM profile")
  while True:
    row = cursor.fetchone()
    if row is None:
      break
    print("id: %s, name: %s, cats: %s" % (row[0], row[1], row[2]))
  print("Number of rows returned: %d" % cursor.rowcount)
  cursor.close()
#@ _FETCHONE_
except mysql.connector.Error as e:
  print("Oops, the statement failed")
  print("Error: %s" % e)

# Note: Following loop would be shorter if written like this:
#  for row in cursor:
# But don't do that because surrounding text in book discusses how to
# use rows as a direct-access array following the fetch operation.

print("Fetch rows with fetchall")
try:
#@ _FETCHALL_
  cursor = conn.cursor()
  cursor.execute("SELECT id, name, cats FROM profile")
  rows = cursor.fetchall()
  for row in rows:
    print("id: %s, name: %s, cats: %s" % (row[0], row[1], row[2]))
  print("Number of rows returned: %d" % cursor.rowcount)
  cursor.close()
#@ _FETCHALL_
except mysql.connector.Error as e:
  print("Oops, the statement failed")
  print("Error: %s" % e)

print("Fetch rows using cursor as iterator")
try:
#@ _CURSOR_ITERATOR_
  cursor = conn.cursor()
  cursor.execute("SELECT id, name, cats FROM profile")
  for (id, name, cats) in cursor:
    print("id: %s, name: %s, cats: %s" % (id, name, cats))
  print("Number of rows returned: %d" % cursor.rowcount)
  cursor.close()
#@ _CURSOR_ITERATOR_
except mysql.connector.Error as e:
  print("Oops, the statement failed")
  print("Error: %s" % e)

print("Execute UPDATE statement (no placeholders)")
try:
#@ _DO_1_
  cursor = conn.cursor()
  cursor.execute("UPDATE profile SET cats = cats+1 WHERE name = 'Sybil'")
  print("Number of rows updated: %d" % cursor.rowcount)
  cursor.close()
  conn.commit()
#@ _DO_1_
except mysql.connector.Error as e:
  print("Oops, the statement failed")
  print("Error: %s" % e)

conn.close()
